internal/godot/artifact/archive/zip.go

```diff
@@ -44,32 +48,114 @@
 
 	defer archive.Close()
 
-	// Extract all files within the archive.
-	for _, f := range archive.File {
-		mode := f.FileInfo().Mode()
-		out := filepath.Join(out, f.Name) //nolint:gosec
-
-		if f.FileInfo().IsDir() {
-			if ctx.Err() != nil {
-				return ctx.Err()
-			}
+	// There doesn't appear to be a good way to read the compressed bytes during
+	// extraction. Instead, use a manual writer and record progress in steps
+	// after each file completes.
+	p, err := newZipProgress(ctx, path)
+	if err != nil {
+		return err
+	}
 
-			if err := os.MkdirAll(out, mode); err != nil {
-				return err
-			}
+	baseDirMode, err := osutil.ModeOf(out)
+	if err != nil {
+		return err
+	}
 
-			continue
+	// Extract all files within the archive.
+	for _, f := range archive.File {
+		if ctx.Err() != nil {
+			return ctx.Err()
 		}
 
-		src, err := archive.Open(f.Name)
-		if err != nil {
+		if err := extractZipFile(ctx, archive, f, out, baseDirMode); err != nil {
 			return err
 		}
 
-		if err := copyFile(ctx, src, mode, out); err != nil {
+		if p != nil {
+			p.Add(f.CompressedSize64)
+		}
+	}
```

Comment:
While this is a valid finding, checking for insecure paths was intentionally left out since the binaries are compiled with `GODEBUG=tarinsecurepath=0,zipinsecurepath=0`. However, it's probably better to check these paths regardless in the event that these functions are used in tests or via a different binary not compiled with those settings.